src/main/scala/tech/cryptonomic/conseil/generic/chain/MultiFetchDecoding.scala

```diff
@@ -0,0 +1,209 @@
+
+}
+
+/** How to use, based on tezos examples
+  *
+  * The current flow to obtain data from tezos is roughly the same, the following flow describes the operations, the expected content
+  * of each step, and an example from tezos
+  *
+  * [Inputs]               -> [FetchJson]                                                                 -> [DecodeJson]
+  * [A list of input data] -> [A list of json data paired with corresponding input, wrapped in an effect] -> [A list of output data, paired with inputs, wrapped in the effect]
+  * [List of offsets]      -> [Future of pairs (offset, json) that describes block data]                  -> [Future of pairs (offset, BlockData) with the block now decoded]
+  * [List of block hashes] -> [Future of pairs (hash, json) that describes operations data]               -> [Future of pairs (hash, List[Operations]) with the ops now decoded]
+  *
+  * which suggest that we may encode this in a trait that separates the steps with a
+  * - fetchJson function, that, given a collection of inputs, returns the effectful result of json data paired with its input value (correlation)
+  * - decodeJson function that will decode every single json value to a full-blown scala data type
+  *
+  * The effectul function is defined in term of a "Kleisli arrow", which is nothing but a fancy-named wrapper for A => F[B], where F[_] is the so-called effect
+  * Using Kleisli enables a plethora of useful methods to call, that will more easily convert a Kleisli arrow to something that accepts different input, or that
+  * adds further operations on input or output, or that adds other effects, like traversing or folding a collection of results if the B result type allows that.
+  *
+  * This enables us to easily compose the `fetchJson` function and `decodeJson` function to a general `fetch` function, that will use the former to retrieve a list of
+  * json values, paired with input, and apply the decoding to each, all the while maintainig the correlation with the original input value.
+  *
+  * In addition to this, we might want to apply multiple decoding of json, and pair the decoded results with the same inputs
+  * Or still we might want to apply more than one fetcher to the same set of inputs, and pair all the results, in a single pass
+  * This is where the `addDecoding` and `typledResults` combinators comes useful.
+  */
+object MultiFetchDecoding {
+  import cats.syntax.foldable._
+  import cats.syntax.functorFilter._
+
+  /* An alias used to define function constraints on the internal dependent types (i.e. `I`, `O`, `E`)
+   * which would be otherwise opaque in any function signature
+   *
+   * Example: the fetcher for tezos blocks might have type `Aux[Future, List, Int, BlockData, String]
+   * where the Int is for the offset from a given block, and String is a representation of a json value.
+   */
+  private type Aux[F[_], T[_], I, O, E] = MultiFetchDecoding[F, T] {
```

Comment:
It would be great to rename `Aux` as well so the code is more friendly to read.

Reply:
I have no trouble with that, but renaming it will lose connection to this technique, which is commonly referred to as the "Aux Pattern"

Reply:
Are the following normal conventions? F for effects, T for containers, I for input, O for output, E for some intermediary representation? If yes, no worries, but some renames might be worth considering, if we are keeping Aux to represent "Aux Pattern"

Reply:
only Aux and F are common, the rest I picked up based on reference to the full word, I input O output E encoded T traversable